### PR TITLE
SAM-3170 Convert Timer() thread to ScheduledThreadPoolExecutor() service

### PR DESCRIPTION
**Pull request by dnhutchins**
This replaces the Timer() that runs a single SubmitTimedAssessmentThread at a fixed rate, with a ScheduledThreadPoolExecutor that uses a pool of threads to execute individual TimedAssessmentRunnables at a fixed rate for each timed assessment in progress.

This change adds the property samigo.timerThreadCount with a default value of 4, in case anyone has a need to tune that value.

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentQueue.java [hunk 1]
```diff
@@ -19,92 +19,101 @@
...
+  // Add a timed assessment to the queue
   public void add(TimedAssessmentGradingModel timedAG){
-	  synchronized (queue) {
-		  queue.put(timedAG.getAssessmentGradingId(), timedAG);
-		  log.debug("***1. TimedAssessmentQueue.add, before schedule timer="+timer);
-		  scheduleTask();
-		  log.debug("***2. TimedAssessmentQueue.add, after schedule timer="+timer);
-	  }
+          // Add grading data to the queue
+          synchronized (queue) {
```
Comment (ern) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentQueue.java]: synchronized on a ConcurentHashMap.put() is not needed as this method is atomic.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentQueue.java [hunk 2]
```diff
@@ -19,92 +19,101 @@
...
-  public boolean isEmpty(){
-    return queue.isEmpty();
+  // Remove a timed assessment from the queue
+  public void remove(long timedAG){
+    log.info("SAMIGO_TIMED_ASSESSMENT:QUEUE:REMOVE ID:" + timedAG);
+    // Stop the task and remove it
+    tasks.get(timedAG).cancel(true);
+    tasks.remove(timedAG);
+    // Remove the grading data from the queue
+    synchronized (queue) {
```
Comment (ern) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentQueue.java]: same as above ConcurrentHashMap.remove() is atomic

**Comment by dnhutchins** [comment]
Removed synchronization from the ConcurrentHashMap put/remove calls.

**Comment by ncaidin** [comment]
CLA is on file. Removed cla? label.

**Comment by jonespm** [comment]
Has one file currently in conflict SubmitTimedAssessmentThread.java

**Comment by dnhutchins** [comment]
This PR removes that file completely. So I'm a but confused as the what the conflict is :/

**Comment by jonespm** [comment]
The change was on https://github.com/sakaiproject/sakai/commit/59e65c39d90823b119051061ae8122dde99f06c2#diff-50bd1651546e802bd2c1022818829dae in SAM-3012

Looks like the diff was changing a event name. So this would have to still be rebased, and the event name EVENT_ASSESSMENT_SUBMITTED_TIMED should be renamed to EVENT_ASSESSMENT_SUBMITTED_TIMER_THREAD

```
-            EventTrackingService.post(EventTrackingService.newEvent(SamigoConstants.EVENT_ASSESSMENT_SUBMITTED_TIMED, notiValues.toString(), siteId, true, SamigoConstants.NOTI_EVENT_ASSESSMENT_TIMED_SUBMITTED));
+            EventTrackingService.post(EventTrackingService.newEvent(SamigoConstants.EVENT_ASSESSMENT_SUBMITTED_TIMER_THREAD, notiValues.toString(), siteId, true, SamigoConstants.NOTI_EVENT_ASSESSMENT_TIMED_SUBMITTED))

```

**Comment by dnhutchins** [comment]
I synced this PR with master, and made the appropriate changes to the new constant.

**Pull request merged**